### PR TITLE
feat: add better error typings

### PR DESCRIPTION
**Pull request by pulkitb2**
close FTL-14880

### CODE REVIEW

File: libs/iota-browser/src/helpers/channel-provider.ts [hunk 1]
```diff
@@ -178,27 +189,44 @@ export class ChannelProvider {
...
+                    getUnexpectedErrorMessage(ErrorCode.SIGNED_REQUEST_EVENT),
+                  )
+                }
+                try {
+                  const claims = jose.decodeJwt(signedRequest.data.jwt)
+                  signedRequestJWT = SignedRequestJWTSchema.parse(claims)
+                } catch (e) {
+                  throw Error(
+                    getUnexpectedErrorMessage(ErrorCode.SIGNED_REQUEST_JWT),
+                  )
```
Comment (aeffinidi) [inline libs/iota-browser/src/helpers/channel-provider.ts:211]: if untilately it's supposed to throw an exception, we could maybe abstract this away into a dedicated function. wdyt ?

File: libs/iota-browser/src/helpers/channel-provider.ts [hunk 2]
```diff
@@ -178,27 +189,44 @@ export class ChannelProvider {
...
 
                 resolve(request)
+              } else if (event.eventType === EventTypes.Error) {
+                let errorEvent: ErrorEvent
+                try {
+                  errorEvent = ErrorEventSchema.parse(event)
+                } catch (e) {
+                  throw Error(getUnexpectedErrorMessage(ErrorCode.ERROR_EVENT))
+                }
+                throw Error(getErrorMessage(errorEvent))
```
Comment (aeffinidi) [inline libs/iota-browser/src/helpers/channel-provider.ts:229]: same comment as above here

File: libs/iota-browser/src/helpers/response-handler.ts [hunk 3]
```diff
@@ -39,16 +48,32 @@ export class ResponseHandler {
...
+                  vpToken = VerifiablePresentationSchema.parse(
+                    JSON.parse(responseCallback.vpToken),
+                  )
+                } catch (e) {
+                  throw Error(
+                    getUnexpectedErrorMessage(
+                      ErrorCode.VERIFIABLE_PRESENTATION_SCHEMA,
+                    ),
+                  )
+                }
```
Comment (aeffinidi) [inline libs/iota-browser/src/helpers/response-handler.ts:76]: same as above

File: libs/iota-browser/src/helpers/response-handler.ts [hunk 4]
```diff
@@ -57,6 +82,14 @@ export class ResponseHandler {
...
                 }
                 resolve(response)
+              } else if (event.eventType === EventTypes.Error) {
+                let errorEvent: ErrorEvent
+                try {
+                  errorEvent = ErrorEventSchema.parse(event)
+                } catch (e) {
+                  throw Error(getUnexpectedErrorMessage(ErrorCode.ERROR_EVENT))
+                }
+                throw Error(getErrorMessage(errorEvent))
```
Comment (aeffinidi) [inline libs/iota-browser/src/helpers/response-handler.ts:92]: same as above

File: libs/iota-browser/src/helpers/channel-provider.ts (hunk 2 above)
Comment (pulkitb2) [reply libs/iota-browser/src/helpers/channel-provider.ts]: addressed

File: libs/iota-browser/src/helpers/response-handler.ts (hunk 4 above)
Comment (pulkitb2) [reply libs/iota-browser/src/helpers/response-handler.ts]: addressed

**Review by rbrazhnyk — APPROVED** [review]

File: libs/iota-browser/src/validators/error.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  'VERIFIABLE_PRESENTATION_SCHEMA' = 'VerifiablePresentationSchema',
+  'ERROR_EVENT' = 'ErrorEvent',
+}
+
+function getIssue(errorEvent: ErrorEvent) {
+  return errorEvent.error.details![0].issue &&
+    errorEvent.error.details![0].issue.length > 0
+    ? errorEvent.error.details![0].issue
+    : errorEvent.error.message
+}
```
Comment (aeffinidi) [inline libs/iota-browser/src/validators/error.ts:16]: isn't this error prone. I'm not sure how we can assume with the `!` that details always exist ?

File: libs/iota-browser/src/validators/error.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+export function getUnexpectedErrorMessage(code: ErrorCode) {
+  return `Unexpected error occured. Error Code: ${code} `
+}
+
+export function getErrorMessage(errorEvent: ErrorEvent): string {
+  return `Something went wrong. ${getIssue(errorEvent)}. Error Code ${errorEvent.error.httpStatusCode}`
+}
```
Comment (aeffinidi) [inline libs/iota-browser/src/validators/error.ts:24]: ok this is a formatter right ? it's not about `getting` something.
What about formatEvenrError ?

File: libs/iota-browser/src/validators/error.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+export function getError(event: ErrorEvent) {
```
Comment (aeffinidi) [inline libs/iota-browser/src/validators/error.ts:26]: can we type the function to and rename it ?
```diff

- export function getError(event: ErrorEvent)
+ export function throwEventParsingError(event: ErrorEvent): never
```

File: libs/iota-browser/src/helpers/channel-provider.ts [hunk 8]
```diff
@@ -176,32 +208,20 @@ export class ChannelProvider {
...
+              reject(e)
+            }
+            if (correlationId !== event.correlationId) {
+              return
+            }
+            if (event.eventType === EventTypes.SignedRequest) {
+              const request = this.getRequest(event)
+              resolve(request)
+            } else if (event.eventType === EventTypes.Error) {
+              getError(event)
```
Comment (aeffinidi) [inline libs/iota-browser/src/helpers/channel-provider.ts:224]: we need to reduce the cognitive complexity here

File: libs/iota-browser/src/validators/error.ts (hunk 5 above)
Comment (pulkitb2) [reply libs/iota-browser/src/validators/error.ts]: If details won't exist, it'll take error.message which will exist and that is expected behaviour.

File: libs/iota-browser/src/helpers/channel-provider.ts (hunk 8 above)
Comment (pulkitb2) [reply libs/iota-browser/src/helpers/channel-provider.ts]: added a research ticket - FTL-15324

**Pull request merged**